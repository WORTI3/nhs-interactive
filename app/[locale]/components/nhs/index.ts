export { default as Button } from './Button';
export { default as Input } from './Input';
export { Header } from './Header';
export { Footer } from './Footer';
export { wrapper as Wrapper } from './wrapper';
export * from './WarningCallout';