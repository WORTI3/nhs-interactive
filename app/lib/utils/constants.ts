export const LOCAL_KEY_ID = "interactive-local-key";
export const UNRELATED_ANSWER = "not-answerable";
export const LOCAL_DB_URL = "http://127.0.0.1:8090";

export enum OpenAIModel {
  DAVINCI_TURBO = "gpt-3.5-turbo",
};