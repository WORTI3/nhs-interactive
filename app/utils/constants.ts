export const LOCAL_KEY_ID = "interactive-local-key";
export const UNRELATED_ANSWER = "not-answerable";
export const LOCAL_DB_URL = "http://127.0.0.1:8090";
export const SERVICE_NAME = "NHS Interactive";
export const SERVICE_DESC = "AI powered search.";
export const SERVICE_URL = "https://nhs-interactive.vercel.app";
export const LOCALES = ['en', 'cy', 'gd'];
export const USE_AI_RESPONSE_KEY = "use-ai-response";

export enum OpenAIModel {
  DAVINCI_TURBO = "gpt-3.5-turbo",
};