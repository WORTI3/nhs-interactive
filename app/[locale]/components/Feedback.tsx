import { Dialog, DialogTitle } from '@headlessui/react';
import { useState } from 'react';
import { Button } from './nhs';
import { useTranslations } from 'next-intl';

const FeedbackDialog = ({ isOpen, onClose, onSubmit }: {isOpen: any, onClose: any, onSubmit: any}) => {
  const translate = useTranslations('search');
  const [comment, setComment] = useState('');
  const [helpful, setHelpful] = useState(false);
  const [error, setError] = useState<string>("");

  const handleSubmit = () => {
    onSubmit({ comment, helpful });
    onClose();
  };

  return (
    <Dialog open={isOpen} onClose={onClose} className="fixed inset-0 overflow-hidden">
      <div className="absolute inset-0 overflow-hidden">
        <div className="absolute inset-0 bg-black opacity-50" />
        <div className="fixed inset-y-0 right-0 max-w-full flex">
          <div className="w-screen max-w-md">
            <div className="h-full bg-white p-4 shadow-xl">
              <DialogTitle>Provide Feedback</DialogTitle>
              <textarea
                className="w-full border rounded mt-4 p-2"
                placeholder="Leave a note..."
                value={comment}
                onChange={(e) => setComment(e.target.value)}
              />
              <div className="mt-4">
                <label>
                  <input
                    type="checkbox"
                    checked={helpful}
                    onChange={(e) => setHelpful(e.target.checked)}
                  />
                  Was this helpful?
                </label>
              </div>
              <div className="mt-4">
                <button
                  className="bg-blue-500 text-white p-2 rounded"
                  onClick={handleSubmit}
                >
                  Submit
                </button>
                <Button
                  id='submit'
                  as='a'
                  data-prevent-double-click
                  onClick={handleSubmit}
                  disabled={!!error && comment === ''}
                >
                  {translate('button.submit')}
                </Button>
                <button
                  className="ml-2 p-2"
                  onClick={onClose}
                >
                  Cancel
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </Dialog>
  );
};

export default FeedbackDialog;
