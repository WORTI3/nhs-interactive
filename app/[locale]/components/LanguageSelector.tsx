import { Fragment } from 'react'
import { Menu, Transition } from '@headlessui/react'
import { IconWorld } from '@tabler/icons-react';
import Link from "next-intl/link";
import { LOCALES } from '../../utils/constants';

function classNames(...classes: any[]) {
  return classes.filter(Boolean).join(' ')
}

export default function LanguageSelector({translate, locale}: {translate: any, locale: string}) {
  return (
    <Menu as='div' className='relative max-lg:absolute max-lg:right-4 float-right inline-block text-left'>
      <div>
        {/* Hover style made need changed to match NHS */}
        <Menu.Button id='language-button' className='inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50'>
          {translate('button.language')}
          <IconWorld
            className='-mr-1 h-5 w-5 text-gray-400'
            aria-hidden='true'
          />
        </Menu.Button>
      </div>

      <Transition
        as={Fragment}
        enter='transition ease-out duration-100'
        enterFrom='transform opacity-0 scale-95'
        enterTo='transform opacity-100 scale-100'
        leave='transition ease-in duration-75'
        leaveFrom='transform opacity-100 scale-100'
        leaveTo='transform opacity-0 scale-95'
      >
        <Menu.Items className='absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none'>
          <div className='langs'>
            {LOCALES.map(lang => {
              if (lang !== locale) {
                return (
                  <Menu.Item key={lang} data-testid="language-item">
                    {({ active }) => (
                      <Link
                        href='/'
                        locale={lang}
                        className={classNames(
                          active ? "bg-gray-100 text-gray-900" : "text-gray-700",
                          "block px-4 py-2 text-sm rounded-md"
                        )}
                      >
                        <p className='mb-0 py-2'>
                          <span
                            className={`mr-4 h-5 w-5 rounded-md fi fis fi-${lang === 'en' ? 'gb-eng' : lang === 'cy' ? 'gb-wls' : 'gb-sct'}`}
                            aria-hidden='true'
                          />
                          {lang === 'en' ? 'English' : lang === 'cy' ? 'Cymraeg' : 'Gaelic'}
                        </p>
                      </Link>
                    )}
                  </Menu.Item>
                );
              }
              return null;
            })}
          </div>
        </Menu.Items>
      </Transition>
    </Menu>
  );
}
