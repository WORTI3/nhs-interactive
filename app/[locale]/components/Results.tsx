import { FC } from "react";
import { UNRELATED_ANSWER } from "../../utils/constants";
import { ResultProps } from "../../utils/interfaces";
import { useApiLog } from "../../utils/hooks";
import { LogData } from "@/types";
import { useTranslations } from "next-intl";

export const Results: FC<ResultProps> = ({ searchQuery, answer, done, onReset }) => {
  const translate = useTranslations('results');
  const logData: LogData = { searchQuery, answer };
  useApiLog(logData, done);
  return (
    <>
      <p className='text-md text-nhs-blue mb-2'>{translate('question')}</p>
      <p className='italic'>{searchQuery.query} (?)</p>
      <hr className='nhsuk-section-break nhsuk-section-break--m nhsuk-section-break--visible' />
      { answer === UNRELATED_ANSWER ? <>
        <p>{translate('error.paragraphOne')}</p>
        <div className='nhsuk-details__text mb-6'>
          <p>{translate('error.paragraphTwo')}</p>
          <ul>
            <li>{translate('error.list.one')}</li>
            <li>{translate('error.list.two')}</li>
            <li>{translate('error.list.three')}</li>
          </ul>
        </div>
        </> : <>
          <p className='text-nhs-blue mb-2'>{translate('answer')}</p>
          <p className='overflow-auto'>
            {answer}
          </p>
        </>
      }
      
      {(done && answer !== UNRELATED_ANSWER) && (
        <>
          <hr className='nhsuk-section-break nhsuk-section-break--m nhsuk-section-break--visible' />
          <p className='text-nhs-blue mb-2'>{translate('sources')}</p>
          {searchQuery.sourceLinks.map((source, index) => (
            <div key={index} className='my-2 overflow-auto max-sm:flex'>
              <a
                className='hover:cursor-pointer hover:underline'
                target='_blank'
                rel='noopener noreferrer'
                href={source}
              >
                <span className="max-sm:flex-1  ">
                  {searchQuery.sourceHeadings[index] ?? 'NHS Search'}
                  <span className="ml-4 max-sm:float-none sm:float-right inline-flex items-center rounded-md bg-sky-50 max-sm:px-1.5 px-2 py-1 max-sm:text-xs text-xs max-sm:font-semibold font-bold text-sky-600 ring-1 ring-inset ring-sky-600/90 hover:bg-sky-100">
                    Verified source: {source.split("//")[1].split("/")[0].replace("www.", "")}
                  </span>
                </span>
              </a>
            </div>
          ))}
          <hr className='nhsuk-section-break nhsuk-section-break--m nhsuk-section-break--visible' />
        </>
      )}
      { (done || answer === UNRELATED_ANSWER) && (
        <button className="nhsuk-button" onClick={onReset}>{translate('button.newQuery')}</button>
      )}
    </>
  );
};
