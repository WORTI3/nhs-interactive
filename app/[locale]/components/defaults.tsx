import { OpenAIModel } from "../../utils/constants";

/**
 * `export const ModelInfo` is defining a React functional component that returns a paragraph element displaying
 * information about a specific OpenAI model. The model information is accessed through the `OpenAIModel` constant defined
 * in the `constants.ts` file. This component can be imported and used in other parts of the codebase.
 * 
 * @function
 * @name ModelInfo
 * @kind variable
 * @returns {JSX.Element}
 * @exports
 */
export const ModelInfo = () => {
  return (
    <p className="nhsuk-body-s nhsuk-u-secondary-text-color nhsuk-u-margin-top-7 nhsuk-u-margin-bottom-0">
      Model: { OpenAIModel.DAVINCI_TURBO }
    </p>
  );
}

/**
 * This code exports a function named `makeSourcesLinks` that takes in two parameters: `answer` and `sourceLinks`. The
 * function splits the `answer` string using a regular expression and maps over the resulting array. If a part of the
 * string matches the regular expression `/\[[0-9]+\]/`, it extracts the number within the square brackets, uses it to
 * retrieve a link from the `sourceLinks` array, and returns an anchor tag with the link and the original part of the
 * string. If the part of the string does not match the regular expression, it returns the original part of the string. The
 * function then returns an array of elements that can be rendered in the UI.
 * 
 * @function
 * @name makeSourcesLinks
 * @kind variable
 * @exports
 */
export const makeSourcesLinks = (answer: string, sourceLinks: string[]) => {
  const elements = answer.split(/(\[[0-9]+\])/).map((part, index) => {
    if (/\[[0-9]+\]/.test(part)) {
      const link = sourceLinks[parseInt(part.replace(/[\[\]]/g, "")) - 1];

      return (
        <a
          key={index}
          className="hover:cursor-pointer text-nhs-blue text-base"
          href={link}
          target="_blank"
          rel="noopener noreferrer"
        >
          {part}
        </a>
      );
    } else {
      return part;
    }
  });

  return elements;
};