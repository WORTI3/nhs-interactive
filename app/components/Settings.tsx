"use client";

import { Button, Input } from "nhsuk-react-components";
import { ChangeEvent, useState } from "react";
import { LOCAL_KEY_ID } from "../lib/utils/constants";
import { useApiKey } from "../lib/utils/hooks";

export const Settings = () => {
  const [showSettings, setShowSettings] = useState<boolean>(false);
  const [apiKey, setApiKey] = useState<string>("");
  const [error, setError] = useState<boolean>(false);

  const handleClear = () => {
    localStorage.removeItem(LOCAL_KEY_ID);
    setApiKey('');
  };

  const handleSave = () => {
    if (apiKey.length !== 51 && apiKey !== "local") {
      setError(true);
      return;
    }
    localStorage.setItem(LOCAL_KEY_ID, JSON.stringify(apiKey));
    setShowSettings(false);
    setError(false);
  };

  useApiKey(setApiKey);
  
  return (
    <>
      <Button disabled={error} as='a' onClick={() => setShowSettings(!showSettings)}>
        {showSettings ? "Hide" : "Show"} Settings
      </Button>
      {showSettings && (
        <>
          <Input
            id="api-key-input"
            label="OpenAi API Key"
            name="api key"
            value={apiKey}
            type="password"
            error={error ? 'provide a valid OpenAi API Key' : ''}
            onChange={(e: ChangeEvent<HTMLInputElement>) => {
              setApiKey(e.target.value);
            }}
          />
          <div className='flex space-x-2 mt-6'>
            <Button secondary onClick={handleSave}>
              Save
            </Button>
            <Button secondary onClick={handleClear}>
              Clear
            </Button>
          </div>
        </>
      )}
    </>
  );
};
