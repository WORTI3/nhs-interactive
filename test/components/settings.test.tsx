import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Settings } from '../../app/components/Settings';
import { LOCAL_KEY_ID } from '../../app/lib/utils/constants';

describe('Settings', () => {
  test('rendered component should match snapshot', () => {
    const { container } = render(<Settings />);
    expect(container).toMatchSnapshot();
  });

  test('should render the component with the correct button text to match snapshot', () => {
    const { container, getByText } = render(<Settings />);
    const showSettingsButton = getByText('Show Settings');

    fireEvent.click(showSettingsButton);

    expect(getByText('Hide Settings')).toBeInTheDocument();
    expect(localStorage.getItem(LOCAL_KEY_ID)).toEqual(null);
    expect(container).toMatchSnapshot();
  });

  test('displays error message when an invalid API key is entered to match snapshot', () => {
    const { container, getByLabelText, getByText } = render(<Settings />);
    const showSettingsButton = getByText('Show Settings');
    fireEvent.click(showSettingsButton); // open setting options

    const apiKeyInput = getByLabelText('OpenAi API Key');
    fireEvent.change(apiKeyInput, { target: { value: 'invalid-api-key' } });
  
    const saveButton = getByText('Save');
    fireEvent.click(saveButton);
  
    expect(apiKeyInput).toHaveValue('invalid-api-key');
    expect(saveButton).toBeInTheDocument();
    expect(getByText('Clear')).toBeInTheDocument();
    expect(getByText('provide a valid OpenAi API Key')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  
  test('clear API key input field when Clear button is used', () => {
    const { getByLabelText, getByText } = render(<Settings />);
    const showSettingsButton = getByText('Show Settings');
    fireEvent.click(showSettingsButton); // open setting options

    const apiKeyInput = getByLabelText('OpenAi API Key');
    fireEvent.change(getByLabelText('OpenAi API Key'), { target: { value: 'invalid-api-key' } });
  
    const clearButton = getByText('Clear');
    fireEvent.click(clearButton);
  
    expect(apiKeyInput).toHaveValue('');
    expect(clearButton).toBeInTheDocument();
    expect(getByText('Save')).toBeInTheDocument();
  });

  test('full user journey should set valid API key into localStorage', () => {
    const { getByLabelText, getByText } = render(<Settings />);
    fireEvent.click(getByText('Show Settings')); // open setting options

    const apiKeyInput = getByLabelText('OpenAi API Key');
    fireEvent.change(apiKeyInput, { target: { value: 'local' } }); // use locally stored api key
  
    const saveButton = getByText('Save');
    fireEvent.click(saveButton);
  
    expect(saveButton).not.toBeInTheDocument();
    expect(getByText('Show Settings')).toBeInTheDocument();
    expect(localStorage.getItem(LOCAL_KEY_ID)).toEqual(JSON.stringify('local'));
  });
  
});