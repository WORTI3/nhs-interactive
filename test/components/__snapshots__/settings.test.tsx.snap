// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings displays error message when an invalid API key is entered to match snapshot 1`] = `
<div>
  <a
    aria-disabled="true"
    class="nhsuk-button nhsuk-button--disabled"
    draggable="false"
    role="button"
  >
    Hide
     Settings
  </a>
  <div
    class="nhsuk-form-group nhsuk-form-group--error"
  >
    <label
      class="nhsuk-label"
      for="api-key-input"
      id="api-key-input--label"
    >
      OpenAi API Key
    </label>
    <span
      class="nhsuk-error-message"
      id="api-key-input--error-message"
      role="alert"
    >
      <span
        class="nhsuk-u-visually-hidden"
      >
        Error: 
      </span>
      provide a valid OpenAi API Key
    </span>
    <input
      aria-labelledby="api-key-input--label"
      class="nhsuk-input nhsuk-input--error"
      id="api-key-input"
      name="api key"
      type="password"
      value="invalid-api-key"
    />
  </div>
  <div
    class="flex space-x-2 mt-6"
  >
    <button
      aria-disabled="false"
      class="nhsuk-button nhsuk-button--secondary"
      type="submit"
    >
      Save
    </button>
    <button
      aria-disabled="false"
      class="nhsuk-button nhsuk-button--secondary"
      type="submit"
    >
      Clear
    </button>
  </div>
</div>
`;

exports[`Settings rendered component should match snapshot 1`] = `
<div>
  <a
    aria-disabled="false"
    class="nhsuk-button"
    draggable="false"
    role="button"
  >
    Show
     Settings
  </a>
</div>
`;

exports[`Settings should render the component with the correct button text to match snapshot 1`] = `
<div>
  <a
    aria-disabled="false"
    class="nhsuk-button"
    draggable="false"
    role="button"
  >
    Hide
     Settings
  </a>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="api-key-input"
      id="api-key-input--label"
    >
      OpenAi API Key
    </label>
    <input
      aria-labelledby="api-key-input--label"
      class="nhsuk-input"
      id="api-key-input"
      name="api key"
      type="password"
      value=""
    />
  </div>
  <div
    class="flex space-x-2 mt-6"
  >
    <button
      aria-disabled="false"
      class="nhsuk-button nhsuk-button--secondary"
      type="submit"
    >
      Save
    </button>
    <button
      aria-disabled="false"
      class="nhsuk-button nhsuk-button--secondary"
      type="submit"
    >
      Clear
    </button>
  </div>
</div>
`;
